demos/shadow/shadow_demo_main/shadow_demo_main.c

```diff
@@ -0,0 +1,644 @@
+ * device shadow. The Device Shadow library does not depend on a MQTT library,
+ * therefore the code for MQTT connections are placed in another file (shadow_demo_helpers.c)
+ * to make it easy to read the code using Device Shadow library.
+ *
+ * This example assumes there is a powerOn state in the device shadow. It does the
+ * following operations:
+ * 1. Establish a MQTT connection by using the helper functions in shadow_demo_helpers.c.
+ * 2. Assemble strings for the MQTT topics of device shadow, by using macros defined by the Device Shadow library.
+ * 3. Subscribe to those MQTT topics by using helper functions in shadow_demo_helpers.c.
+ * 4. Publish a desired state of powerOn by using helper functions in shadow_demo_helpers.c.  That will cause
+ * a delta message to be sent to device.
+ * 5. Handle incoming MQTT messages in eventCallback, determine whether the message is related to the device
+ * shadow by using a function defined by the Device Shadow library (Shadow_MatchTopic). If the message is a
+ * device shadow delta message, set a flag for the main function to know, then the main function will publish
+ * a second message to update the reported state of powerOn.
+ * 6. Handle incoming message again in eventCallback. If the message is from update/accepted, verify that it
+ * has the same clientToken as previously published in the update message. That will mark the end of the demo.
+ */
+
+/* Standard includes. */
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* POSIX includes. */
+#include <unistd.h>
+
+#include "shadow_config.h"
+
+/* SHADOW API header. */
+#include "shadow.h"
+
+/* JSON API header. */
+#include "json.h"
+
+/* shadow demo helpers header. */
+#include "shadow_demo_helpers.h"
+
+
+/**
+ * @brief Format string representing a Shadow document with a "desired" state.
+ *
+ * The real json document will look like this:
+ * {
+ *   "state": {
+ *     "desired": {
+ *       "powerOn": 1
+ *     }
+ *   },
+ *   "clientToken": "021909"
+ * }
+ *
+ * Note the client token, which is required for all Shadow updates. The client
+ * token must be unique at any given time, but may be reused once the update is
+ * completed. For this demo, a timestamp is used for a client token.
+ */
+#define SHADOW_DESIRED_JSON     \
+    "{"                         \
+    "\"state\":{"               \
+    "\"desired\":{"             \
+    "\"powerOn\":%01d"          \
+    "}"                         \
+    "},"                        \
+    "\"clientToken\":\"%06lu\"" \
+    "}"
+
+/**
+ * @brief The expected size of #SHADOW_DESIRED_JSON.
+ *
+ * Because all the format specifiers in #SHADOW_DESIRED_JSON include a length,
+ * its full size is known at compile-time. "In your own application, this might
+ * not be true, then you should calculate the size of the json doc at run time.
+ */
+#define SHADOW_DESIRED_JSON_LENGTH    ( sizeof( SHADOW_DESIRED_JSON ) - 3 )
```

Comment:
This doesn't make sense to me.  sizeof is not interpreting format specifiers, or am I crazy?

Reply:
The result is the pre-calculated and fixed. The reason why sizeof( SHADOW_DESIRED_JSON ) need to minus 3 is:
1. the length of "%01d" is 4.
2. the length of %06lu is 5.
3. the actual length we will use in case 1. is 1. (the powerOn state is either 1 or 0).
4. the actual length we will use in case 2. is 6. 
5. thus the additional size 3 = 4 + 5 - 1 - 6 + 1 (termination character).

Reply:
OK, that's a good suggestion, will add it.

Reply:
[This commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1104/commits/1c5938bf39c086c340dcc4e83dedc5c2482cf7a7) add more explanation in the comment.